platform/darwin/src/NSPredicate+MGLAdditions.mm

```diff
@@ -294,13 +294,39 @@ + (instancetype)mgl_predicateWithJSONObject:(id)object {
         return [NSPredicate predicateWithValue:YES];
     }
     if ([op isEqualToString:@"all"]) {
-        NSArray *subpredicates = MGLSubpredicatesWithJSONObjects([objects subarrayWithRange:NSMakeRange(1, objects.count - 1)]);
+        NSArray *jsonObjects = [objects subarrayWithRange:NSMakeRange(1, objects.count - 1)];
+        NSArray *subpredicates = MGLSubpredicatesWithJSONObjects(jsonObjects);
+        if (jsonObjects.count == 2) {
+            // Determine if the expression is of BETWEEN type
+            if ([jsonObjects[0] isKindOfClass:[NSArray class]] &&
+                [jsonObjects[1] isKindOfClass:[NSArray class]]) {
+                NSArray *leftCondition = jsonObjects[0];
+                NSArray *rightCondition = jsonObjects[1];
+                NSString *leftOperator = leftCondition.firstObject;
+                NSString *rightOperator = rightCondition.firstObject;
+                if ([leftOperator isEqualToString:@"<="] && [rightOperator isEqualToString:@"<="]) {
+                    return [NSPredicate predicateWithFormat:@"%@ BETWEEN %@", [NSExpression mgl_expressionWithJSONObject:leftCondition[2]], @[[NSExpression mgl_expressionWithJSONObject:leftCondition[1]], [NSExpression mgl_expressionWithJSONObject:rightCondition[2]]]];
+                } else if([leftOperator isEqualToString:@">="] && [rightOperator isEqualToString:@"<="]) {
+                    return [NSPredicate predicateWithFormat:@"%@ BETWEEN %@", [NSExpression mgl_expressionWithJSONObject:leftCondition[1]], @[[NSExpression mgl_expressionWithJSONObject:leftCondition[2]], [NSExpression mgl_expressionWithJSONObject:rightCondition[2]]]];
+                } else if([leftOperator isEqualToString:@">="] && [rightOperator isEqualToString:@">="]) {
+                    return [NSPredicate predicateWithFormat:@"%@ BETWEEN %@", [NSExpression mgl_expressionWithJSONObject:leftCondition[1]], @[[NSExpression mgl_expressionWithJSONObject:leftCondition[2]], [NSExpression mgl_expressionWithJSONObject:rightCondition[1]]]];
+                }
+            }
+        }
         return [NSCompoundPredicate andPredicateWithSubpredicates:subpredicates];
     }
     if ([op isEqualToString:@"any"]) {
         NSArray *subpredicates = MGLSubpredicatesWithJSONObjects([objects subarrayWithRange:NSMakeRange(1, objects.count - 1)]);
         return [NSCompoundPredicate orPredicateWithSubpredicates:subpredicates];
     }
+    if ([op isEqualToString:@"match"]) {
+        NSArray *subpredicates = MGLSubexpressionsWithJSONObjects([objects subarrayWithRange:NSMakeRange(1, objects.count - 1)]);
+        return [NSPredicate predicateWithFormat:@"%@ IN %@" argumentArray:subpredicates];
```

Comment:
The following JSON objects:

```json
[
  "match",
  ["get", "a"],
  1, true,
  2, true,
  3, true,
  4, true,
  false
]
```

```json
[
  "match",
  ["get", "a"],
  [1, 2, 3, 4], true,
  false
]
```

should translate to an NSExpression with the format string `a IN {1, 2, 3, 4}`. We don’t currently support translating an NSExpression to a `match` that consolidates cases like that (#11539), but I think it’s possible for us to encounter such an expression here if the style JSON defines it that way.

Not all `match` expressions can be represented by the `IN` operator, only expressions that have two unique output values, `true` and `false`.